fvm/environment/uuids_test.go

```diff
@@ -328,15 +329,15 @@ func TestUUIDGeneratorHardcodedPartitionIdGeneration(t *testing.T) {
 
 	value, err = uuids.GenerateUUID()
 	require.NoError(t, err)
-	require.Equal(t, value, uint64(0xdefffffffffffffd))
+	require.Equal(t, value, uint64(0xffffdefffffffffd))
 
 	value, err = uuids.getUint64()
 	require.NoError(t, err)
 	require.Equal(t, value, maxUint56Minus2+1)
 
 	value, err = uuids.GenerateUUID()
 	require.NoError(t, err)
-	require.Equal(t, value, uint64(0xdefffffffffffffe))
+	require.Equal(t, value, uint64(0xffffdefffffffffe))
 
```

Comment:
I'm not sure about the implementation.

We know that the UUID generator generates UUID for the same partition, and the partition is determined by the block ID and transaction index. 

In other words, given a UUID Generator, if we keep calling `GenerateUUID`, the returned value always the same byte at the partition byte position, which is the 3rd byte. 

However, L201 takes the stored uuid value and add `1` to it:
```
	err = generator.setUint56(value + 1)
```

This works if partition is the 1st byte, but won't work if partition is 3rd byte. 
See this case: if now the UUID for partition 2 is: `3298534883327` (hex: `00 00 02 FF FF FF FF FF`)
```
C7 C6  P C5 C4 C3 C2 C1
00 00 02 FF FF FF FF FF
```
Then calling `GenerateUUID()` again, will call `setUint56(value + 1)`, which becomes `00 00 03 00 00 00 00 00`, which would not increment `C6`, and causing the counter to go back to 0 again.

We need to have a test case, for a UUID generator for partition 2 and it's value is `00 00 02 FF FF FF FF FF`, the next two UUIDs should be `00 01 02 00 00 00 00 00`, and  `00 01 02 00 00 00 00 01`.

Reply:
What happens is that we first call `generator.getUint64()` which gets the stored counter for that partition (`00 00 00 FF FF FF FF FF`). then stores it back  as incremented by 1  `generator.setUint56(value + 1)` ( 00 00 01 00 00 00 00 00). Only after this is the value cut up and the partition bytes inserted.

value: `00 00 00 FF FF FF FF FF` 
partition:  `02`
```
(value & 0xFF_FF00_0000_0000) << 8) -> 00 00 00 00 00 00 00 00
(partition << 40)                   -> 00 00 02 00 00 00 00 00
(value & 0xFF_FFFF_FFFF)            -> 00 00 00 FF FF FF FF FF
                                       00 00 02 FF FF FF FF FF
```

and the next one:
value: `00 00 01 00 00 00 00 00` 
partition:  `02`
```
(value & 0xFF_FF00_0000_0000) << 8) -> 00 01 00 00 00 00 00 00
(partition << 40)                   -> 00 00 02 00 00 00 00 00
(value & 0xFF_FFFF_FFFF)            -> 00 00 00 00 00 00 00 00
                                       00 01 02 00 00 00 00 00
```

I do agree that a test for this continuation would be great!

Reply:
sure, of course, I missed that part of the comment

Reply:
Oh, I knew where I got confused. I thought the stored register value contains the partition and the counter, but actually it is just the counter. The register ID contains the partition already.

Reply:
yeah naming of the functions are confusing, `setUint56` and `getUint64` makes you feel like you are setting 56 bits of 64 bit int, and then getting it as 64 bit with partition.

Reply:
exactly